<a href="https://colab.research.google.com/github/Saulo-Ferraz/PythonCodes/blob/main/Lista_7_Saulo_Raphael.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#Create a virtual env: py -3 -m venv .venv
#Activate virtual env: .venv\scripts\activate
#Install libraries on virtual env: python -m pip install matplotlib

In [ ]:
import pandas as pd
import numpy as np
import os

Para responder os exercícios a seguir, utilize os dados fornecidos na planilha "Bilateral_Country_Shares_govt_nationality_full.csv"

In [ ]:
shares_gov = pd.read_csv('/content/Bilateral_Country_Shares_govt_nationality_full (1).csv')
shares_gov.head(5)

year investor_name investor  issuer_nationality_name issuer_nationality  \
0  2017     Australia      AUS                Argentina                ARG   
1  2017     Australia      AUS                Australia                AUS   
2  2017     Australia      AUS                   Brazil                BRA   
3  2017     Australia      AUS                   Canada                CAN   
4  2017     Australia      AUS  European Monetary Union                EMU   

   portfolio_share  
0         0.009804  
1         0.278169  
2         0.011028  
3         0.042183  
4         0.140563

In [ ]:
shares_gov.columns.values

array(['year', 'investor_name', 'investor', 'issuer_nationality_name',
       'issuer_nationality', 'portfolio_share'], dtype=object)

1) Retorne em ordem crescente, os países que contem maior "portfolio_share" do Brasil nos anos de 2006, 2009 e 2013. Apresente um dataframe para cada ano mencionado (ou seja, 3 dataframes - 06', 09', 13'), contendo duas colunas: "Nome do País" e "Porfolio Share"

In [ ]:
db_issuer = shares_gov[shares_gov["issuer_nationality"]=="BRA"].query("year in (2006,2009, 2013)")
db_2013 = db_issuer[db_issuer['year']==2013].sort_values('portfolio_share',ascending = False)
db_2009 = db_issuer[db_issuer['year']==2009].sort_values('portfolio_share',ascending = False)
db_2006 = db_issuer[db_issuer['year']==2006].sort_values('portfolio_share',ascending = False)

print("DataFrame 2006: \n")
print(db_2006.rename(columns={'investor_name': 'Nome do País', 'portfolio_share': 'Portfolio Share'})[['Nome do País', 'Portfolio Share']])

print("\nDataFrame 2009: \n")
print(db_2009.rename(columns={'investor_name': 'Nome do País', 'portfolio_share': 'Portfolio Share'})[['Nome do País', 'Portfolio Share']])

print("\nDataFrame 2013: \n")
print(db_2013.rename(columns={'investor_name': 'Nome do País', 'portfolio_share': 'Portfolio Share'})[['Nome do País', 'Portfolio Share']])

DataFrame 2006: 

                 Nome do País  Portfolio Share
2112                  Denmark         0.052369
2135  European Monetary Union         0.004932
2222            United States         0.003399
2206           United Kingdom         0.003057
2097                   Canada         0.000880

DataFrame 2009: 

                 Nome do País  Portfolio Share
1635                  Denmark         0.050537
1743           United Kingdom         0.007333
1663  European Monetary Union         0.006835
1599                Australia         0.005724
1763            United States         0.004142
1617                   Canada         0.001473

DataFrame 2013: 

                Nome do País  Portfolio Share
836                  Denmark         0.046965
866  European Monetary Union         0.015231
952           United Kingdom         0.007407
972            United States         0.005876
811                   Canada         0.004780


2) É usual que os países se financiem por emissão de dívida interna, ou seja, que o país contenha um "porfolio_share" considerável dos seus próprios títulos relativo aos títulos de outros países. Exemplo, em 2017, ~ 27.8% da dívida nacional australiana era composta de títulos emitidos pelo próprio governo australiano. Indique no ano de 2014, qual dos países analisados possuía maior dívida interna?

In [ ]:
db_2014 = shares_gov[shares_gov["year"]==2014]
db_self_debt = db_2014[db_2014["investor"] == db_2014["issuer_nationality"]].sort_values(by='portfolio_share', ascending=True).head(1)

print("País com maior dívida interna em 2014: \n")
print(db_self_debt[['investor', 'portfolio_share','issuer_nationality_name']])

País com maior dívida interna em 2014: 

    investor  portfolio_share issuer_nationality_name
737      CHE         0.272654             Switzerland


Para responder os exercícios a seguir, utilize os dados fornecidos na planilha "236_data.xlsx" (na aba "Master")

In [ ]:
currencies = pd.read_excel("/content/236_data.xlsx", sheet_name= "Master", skiprows = 7, header = 0)

currencies.columns.values

array(['Country', 'Afghanistan', 'Albania', 'Algeria', 'Andorra',
       'Angola', 'Anguilla', 'Antigua and Barbuda', 'Argentina',
       'Armenia', 'Aruba', 'Australia', 'Austria', 'Azerbaijan',
       'Bahamas', 'Bahrain', 'Bangladesh', 'Barbados', 'Belarus',
       'Belgium', 'Belize', 'Benin', 'Bermuda', 'Bhutan', 'Bolivia',
       'Bosnia and Herzegovina', 'Botswana', 'Brazil',
       'Brunei Darussalam', 'Bulgaria', 'Burkina Faso', 'Burundi',
       'Cabo Verde', 'Cambodia', 'Cameroon', 'Canada',
       'Central African Republic', 'Chad', 'Chile', 'China', 'Colombia',
       'Comoros', 'Congo, Democratic Republic of', 'Congo, Republic of',
       'Costa Rica', "Cote d'Ivoire", 'Croatia', 'Curacao & St. Maarten',
       'Cyprus', 'Czech Republic', 'Denmark', 'Djibouti', 'Dominica',
       'Dominican Republic', 'Ecuador', 'Egypt', 'El Salvador',
       'Equatorial Guinea', 'Eritrea', 'Estonia', 'Ethiopia', 'Fiji',
       'Finland', 'France', 'Gabon', 'Gambia', 'Georgia', 'Germany',

3) Qual o ano com maior frequência de "Freely_falling" como moeda de referência?

In [ ]:
year_col = currencies.iloc[:,0]
count_ff = currencies.apply(pd.Series.value_counts, axis=1)[['Freely_falling']].fillna(0)
final_count_ff = pd.concat([year_col, count_ff], axis=1).sort_values(by = 'Freely_falling', ascending = False)
final_count_ff.iloc[0,0]

1993

4) Com a queda da União Soviética, além da separação política, foi possível observar uma movimentação na moeda utilizada como âncora. Retorne um dataframe que contenha o ano em que cada país da lista abaixo deixou de utilizar o RUB (ruble)

In [130]:
#Para ajudar, aqui está uma lista com os nomes dos países (ex-Russia) que deve ser utilizada p este exercício
urss_list = ['Ukraine', 'Georgia', 'Uzbekistan', 'Armenia', 'Azerbaijan', 'Kazakhstan', 'Kyrgyzstan', 'Moldova',
'Turkmenistan', 'Tajikistan', 'Latvia', 'Lithuania', 'Estonia']

for country in urss_list[:]:
    if country in currencies.columns:
        country_rub_years = currencies[currencies[country] == 'RUB']['Country']
        last_year = country_rub_years.iloc[-1]
        print(f'País: {country}')
        if not country_rub_years.empty:
            print(last_year)

País: Ukraine
1990
País: Georgia
1991
País: Uzbekistan
1991
País: Armenia
1991
País: Azerbaijan
1992
País: Kazakhstan
1990
País: Kyrgyzstan
1991
País: Moldova
1990
País: Turkmenistan
1992
País: Tajikistan
1991
País: Latvia
1990
País: Lithuania
1990
País: Estonia
1990


In [131]:
selected_df = currencies[currencies.columns.intersection(urss_list)]
selected_df_years = pd.concat([year_col, selected_df], axis=1).set_index("Country")

years = []
for country in urss_list:
    selected_df_years[country] = selected_df_years[country].ne(selected_df_years[country].shift().bfill()).astype(int)
    year_y = selected_df_years[country].idxmax()
    years.append(year_y)

final_df = pd.DataFrame({"Country":urss_list, "Years" : years})



5) Retorne a década em que houve maior troca de "Currency Anchors" no mundo

In [132]:
currencies_indexed = currencies.set_index('Country')
countries_list = list(currencies.columns)[1:199]

for country in countries_list:
    currencies_indexed[country] = currencies_indexed[country].ne(currencies_indexed[country].shift().bfill()).astype(int)

currencies_indexed['change'] = currencies_indexed.sum(axis=1)
year_x = currencies_indexed['change'].sort_values(ascending=False).reset_index().iloc[0,0]
year_x

<ipython-input-132-d06898cdff67>:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  currencies_indexed['change'] = currencies_indexed.sum(axis=1)


1999